### PR TITLE
[contributing] Add a tools guide. 

### PR DESCRIPTION
**Pull request by jverkoey**
This document describes recommended tools for the team.

### CODE REVIEW

**Comment by yarneo** [comment]
I personally disagree that you will need so much tooling to start contributing to our codebase. these tools should be convenient add-ons but not be core to contribution.

**Comment by jverkoey** [comment]
Made the documentation focus on core contributors.

**Comment by jverkoey** [comment]
> These tools should be convenient add-ons but not be core to contribution.

I disagree. I expect the team to operate as a unit when it comes to our workflow. Using these tools translates to less cognitive load over time (e.g. style nits or ensuring that people are putting the full PR description in their commit messages are minimized for the most common cases).

**Comment by codeman7** [comment]
So we would run `git clang-format` before pushing any PR?

**Comment by yarneo** [comment]
If this isn't for any contributor but for the team contributors, I would be more explicit about that. "core" seems like an arbitrary word for me. Maybe elaborate on what a core contributor is would make it easier. As for having to use clang-format for every code we push, I am still partial about that. I feel like when we use clang-format it is mostly a fixer-upper but we never learn from it on what it is to write good code, and we start heavily relying on a bot to fix our stuff. But that's for a separate discussion.

**Comment by romoore** [comment]
I think as long as the code conforms to the Objective-C style guide (and other internal requirements), which tool is used shouldn't matter. I like that there's a tool available to make it easier, but since it still requires manual review (it isn't a perfect tool), it shouldn't be required.

File: contributing/core-contributor-getting-started.md
Comment (romoore) [inline contributing/core-contributor-getting-started.md]: I haven't installed this and I have no problem writing commit messages. In some cases, the summary/description isn't updated and so I've gone and fixed it when I had to manually copy-paste. 

I think these are better listed as optionals/helpful tools. Having them as a "checklist" makes them feel required.

File: contributing/core-contributor-getting-started.md
Comment (romoore) [inline contributing/core-contributor-getting-started.md]: Is this secure? Do we need to check against a known hash or something?

Comment (jverkoey) [reply contributing/core-contributor-getting-started.md]: Done.

**Review by yarneo — COMMENTED** [review]
Who are the core contributors? Is it just the team, team + Google contributors? Is it people who contribute a lot in general? Is this a known terminology in our docs?

**Comment by jverkoey** [comment]
> As for having to use clang-format for every code we push, I am still partial about that.

As discussed in the team meeting this week we will start enforcing code style using a clang-format presubmit as part of our pull requests. This will identify style issues that could be resolved by running clang-format. If someone prefers to modify the style by hand I don't think we can stop anyone, but I think that using a tool to auto-format will be much easier.

**Comment by jverkoey** [comment]
> Who are the core contributors? Is it just the team, team + Google contributors? Is it people who contribute a lot in general? Is this a known terminology in our docs?

Renamed the document to "tools".

**Comment by jverkoey** [comment]
> I think as long as the code conforms to the Objective-C style guide (and other internal requirements), which tool is used shouldn't matter. I like that there's a tool available to make it easier, but since it still requires manual review (it isn't a perfect tool), it shouldn't be required.

+1

**Comment by yarneo** [comment]
I don't mind having clang-format in the presubmits. I do think everyone is entitled to use whatever way they wish to produce successful code. If it tells me I did something wrong, I want to know why. I want to *know* the style guide, and not just have a quick fix.

**Comment by jverkoey** [comment]
> If it tells me I did something wrong, I want to know why. I want to know the style guide, and not just have a quick fix.

+1 the presubmit will describe changes it expects to be made in the presubmit logs.

**Review by yarneo — APPROVED** [review]

**Review by codeman7 — APPROVED** [review]

**Comment by jverkoey** [comment]
Note: this PR requires LGTM from all reviewers.

File: contributing/tools.md
Comment (ianegordon) [inline contributing/tools.md]: s/git clang-format -h/clang-format -h/

**Review by ianegordon — APPROVED** [review]
Please consider comment before merge.
format all the things!

**Comment by jverkoey** [comment]
Waiting on approval from the following folks:

@andrewoverton andrewoverton
 @galiak11 galiak11
 @randallli randallli
 @wenyuzhang666 wenyuzhang666
 @romoore romoore

**Review by andrewoverton — APPROVED** [review]

File: contributing/tools.md
Comment (romoore) [inline contributing/tools.md]: Let's make the header more human-focused.  Suggestion: "Automatically format your Git commit with clang-format"

File: contributing/tools.md
Comment (romoore) [inline contributing/tools.md]: I believe the team should be using the internal version, yes? Should we install using our internal package manager?

File: contributing/tools.md
Comment (romoore) [inline contributing/tools.md]: If this succeeds, are we done or do we continue to the next step?

**Review by romoore — APPROVED** [review]

**Comment by jverkoey** [comment]
Waiting for feedback from:

@galiak11 galiak11
 @randallli randallli
 @wenyuzhang666 wenyuzhang666

File: contributing/tools.md
Comment (jverkoey) [reply contributing/tools.md]: Added a comment to clarify.

**Review by randallli — APPROVED** [review]
shall we add it to 
https://github.com/material-components/material-components-ios/blob/develop/scripts/install_contributor_tools?

**Comment by jverkoey** [comment]
I think I may remove that tool because it's often wrong and overly-assumptive about how and where someone wants to install their software.

**Comment by jverkoey** [comment]
https://github.com/material-components/material-components-ios/issues/5016

**Review by fluxminds — APPROVED** [review]

File: contributing/tools.md
Comment (galiak11) [inline contributing/tools.md:36]: Not sure what "from origin/develop to HEAD" means. Is it running the clang formatter on files currently in my working directory?

Comment (galiak11) [reply contributing/tools.md:36]: Actually. I read the "git-clang-format" file so I get it now. 

Perhaps we can add text suggesting to read the file to see more command options?

**Review by galiak11 — APPROVED** [review]

**Comment by yarneo** [comment]
@jverkoey is this ready to be merged?

File: contributing/tools.md
Comment (ajsecord) [inline contributing/tools.md:8]: It's a little odd to put a script in docs...why not just add this to `install_contributor_tools`? I feel like we've gone a complete circle. :D

**Review by ajsecord — CHANGES_REQUESTED** [review]

File: contributing/tools.md
Comment (jverkoey) [reply contributing/tools.md:8]: I think we’re going to move away from the automated installer because there are enough cases where folks have already installed their own dependencies and our scripts fail to consider these edge cases.

**Review by ajsecord — APPROVED** [review]

**Pull request merged**
